### PR TITLE
[SPARK-15445][SQL] Build fails for java 1.7 after adding java.mathBigInteger support

### PR DESCRIPTION
**Pull request by techaddict**
## What changes were proposed in this pull request?

Using longValue() and then checking whether the value is in the range for a long manually.
## How was this patch tested?

Existing tests

### CODE REVIEW

**Comment by techaddict** [comment]
cc: @srowen

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala [hunk 1]
```diff
@@ -134,7 +135,14 @@ final class Decimal extends Ordered[Decimal] with Serializable {
   def set(bigintval: BigInteger): Decimal = {
     try {
       this.decimalVal = null
-      this.longVal = bigintval.longValueExact()
+      // TODO: Remove this once we migrate to java8 and use longValueExact() instead.
+      val isExactLong = bigintval.compareTo(BigInteger.valueOf(JLong.MAX_VALUE)) <= 0 &&
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala]: Maybe extract constants for a `BigInteger` representation of the max/min values rather than compute each time? Also is this all uses of java.util.Long that can be JLong?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala [hunk 2]
```diff
@@ -134,7 +135,14 @@ final class Decimal extends Ordered[Decimal] with Serializable {
...
+        bigintval.compareTo(BigInteger.valueOf(JLong.MIN_VALUE)) >= 0
+      if(isExactLong) {
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala]: Space after if. Don't throw an exception just to catch it; throw the exception below here. No need for "else"

Comment (techaddict) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala]: sorry didn't get the part throw the exception below here.

Comment (techaddict) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala]: Since we don't need the try catch, this would be much better
```scala
def set(bigintval: BigInteger): Decimal = {
    this.decimalVal = null
    // TODO: Remove this once we migrate to java8 and use longValueExact() instead.
    val isExactLong = bigintval.compareTo(LONG_MAX_BIG_INT) <= 0 &&
      bigintval.compareTo(LONG_MIN_BIG_INT) >= 0
    if (isExactLong) {
      this.longVal = bigintval.longValue()
    } else {
      throw new IllegalArgumentException(s"BigInteger ${bigintval} too large for decimal")
    }
    this._precision = DecimalType.MAX_PRECISION
    this._scale = 0
    this
  }
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala [hunk 3]
```diff
@@ -174,7 +176,7 @@ final class Decimal extends Ordered[Decimal] with Serializable {
 
   def toScalaBigInt: BigInt = BigInt(toLong)
 
-  def toJavaBigInteger: java.math.BigInteger = java.math.BigInteger.valueOf(toLong)
+  def toJavaBigInteger: BigInteger = BigInteger.valueOf(toLong)
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala]: Although I support removing the redundant qualified name, I wonder if we should also un-qualify scala.math.BigInt, or else leave both qualified even though they're not ambiguous. BigDecimal is used unqualified while the Java version is always qualified, to disambiguate. I guess it would become consistent if BigInt were not qualified.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala [hunk 4]
```diff
@@ -132,17 +133,18 @@ final class Decimal extends Ordered[Decimal] with Serializable {
...
-      this._scale = 0
-      this
+    this.decimalVal = null
+    // TODO: Remove this once we migrate to java8 and use longValueExact() instead.
+    val isExactLong = bigintval.compareTo(LONG_MAX_BIG_INT) <= 0 &&
+      bigintval.compareTo(LONG_MIN_BIG_INT) >= 0
+    if (isExactLong) {
+      this.longVal = bigintval.longValue()
+    } else {
+      throw new IllegalArgumentException(s"BigInteger ${bigintval} too large for decimal")
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala]: How about just 

```
require(bigintval.compareTo(LONG_MAX_BIG_INT) <= 0 && bigintval.compareTo(LONG_MIN_BIG_INT) >= 0, ...message...)
this.longValue = ...
```

**Comment by techaddict** [comment]
@srowen All tests passed.

**Comment by gatorsmile** [comment]
cc @kevinyu98

**Comment by kevinyu98** [comment]
@techaddict @srowen @cloud-fan @gatorsmile : Hi Sandeep , thanks for fixing this. I didn't realize the method is java 1.8 only. The code looks good to me.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala [hunk 5]
```diff
@@ -412,8 +414,8 @@ object Decimal {
     value match {
       case j: java.math.BigDecimal => apply(j)
       case d: BigDecimal => apply(d)
-      case k: scala.math.BigInt => apply(k)
-      case l: java.math.BigInteger => apply(l)
+      case k: BigInt => apply(k)
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala]: this is too confusing. please keep the qualifiers. i'd actually keep qualifiers everywhere.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala [hunk 6]
```diff
@@ -132,17 +133,15 @@ final class Decimal extends Ordered[Decimal] with Serializable {
...
-      this._precision = DecimalType.MAX_PRECISION
-      this._scale = 0
-      this
-    }
-    catch {
-      case e: ArithmeticException =>
-        throw new IllegalArgumentException(s"BigInteger ${bigintval} too large for decimal")
-     }
+    // TODO: Remove this once we migrate to java8 and use longValueExact() instead.
+    require(
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala]: why doing this as a require intsead of try catch?

**Comment by rxin** [comment]
@techaddict thanks for doing this. Can you be more surgical in fixes like this? i.e. just fix the line that is causing problem.

**Comment by rxin** [comment]
Also I think you can remove "HOTFIX" -- those are for master build breaking changes, which isn't happening here yet. (fwiw we don't run java 7 on master)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala (hunk 6 above)
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala]: Now the error is not detected by catching an exception but by manually checking the value. Require is the simplest  way to get the desired result which is an IAE.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala (hunk 5 above)
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala]: The code is inconsistent in this regard though. I thought it worth cleaning up but don't feel strongly about it. Alternative is to qualify both classes everywhere consistently

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala (hunk 6 above)
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala]: Doesn't this have a potential perf hit? Also - is this actually related to the problem this ticket is solving?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala (hunk 5 above)
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala]: I agree it's better to make it consistent, but let's not mix it into a simple fix for java 7 ...

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala (hunk 6 above)
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala]: The problem is we can't call a method that will throw an exception if the value is out of bounds during conversion. We just check it manually. I actually expect this is faster but it is vanishingly small as a difference. I don't think there is a choice  if Java 7 is being supported.

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala]: Ah ok. Then well - we have to do this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala [hunk 7]
```diff
@@ -382,6 +381,9 @@ object Decimal {
   private[sql] val ZERO = Decimal(0)
   private[sql] val ONE = Decimal(1)
 
+  private[types] val LONG_MAX_BIG_INT = BigInteger.valueOf(JLong.MAX_VALUE)
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala]: this can just be private

**Comment by techaddict** [comment]
@rxin @srowen I've removed qualifiers

**Comment by srowen** [comment]
Merged to master/2.0
